core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1255,27 +1255,46 @@ class DAGScheduler(
-              s"has failed the maximum allowable number of " +
-              s"times: ${Stage.MAX_CONSECUTIVE_FETCH_FAILURES}. " +
-              s"Most recent failure reason: ${failureMessage}", None)
-          } else {
-            if (failedStages.isEmpty) {
-              // Don't schedule an event to resubmit failed stages if failed isn't empty, because
-              // in that case the event will already have been scheduled.
-              // TODO: Cancel running tasks in the stage
-              logInfo(s"Resubmitting $mapStage (${mapStage.name}) and " +
-                s"$failedStage (${failedStage.name}) due to fetch failure")
-              messageScheduler.schedule(new Runnable {
-                override def run(): Unit = eventProcessLoop.post(ResubmitFailedStages)
-              }, DAGScheduler.RESUBMIT_TIMEOUT, TimeUnit.MILLISECONDS)
+          val shouldAbortStage =
+            failedStage.failedOnFetchAndShouldAbort(task.stageAttemptId) ||
+            disallowStageRetryForTest
+
+          if (shouldAbortStage) {
+            val abortMessage = if (disallowStageRetryForTest) {
+              "Fetch failure will not retry stage due to testing config"
+            } else {
+              s"""$failedStage (${failedStage.name})
+                 |has failed the maximum allowable number of
+                 |times: ${Stage.MAX_CONSECUTIVE_FETCH_FAILURES}.
+                 |Most recent failure reason: $failureMessage""".stripMargin.replaceAll("\n", " ")
             }
+            abortStage(failedStage, abortMessage, None)
+          } else { // update failedStages and make sure a ResubmitFailedStages event is enqueued
+            // TODO: Cancel running tasks in the failed stage -- cf. SPARK-17064
+            val noResubmitEnqueued = !failedStages.contains(failedStage)
             failedStages += failedStage
             failedStages += mapStage
+            if (noResubmitEnqueued) {
+              // We expect one executor failure to trigger many FetchFailures in rapid succession,
+              // but all of those task failures can typically be handled by a single resubmission of
+              // the failed stage.  We avoid flooding the scheduler's event queue with resubmit
+              // messages by checking whether a resubmit is already in the event queue for the
+              // failed stage.  If there is already a resubmit enqueued for a different failed
+              // stage, that event would also be sufficient to handle the current failed stage, but
+              // producing a resubmit for each failed stage makes debugging and logging a little
+              // simpler while not producing an overwhelming number of scheduler events.
+              logInfo(
+                s"Resubmitting $mapStage (${mapStage.name}) and " +
+                s"$failedStage (${failedStage.name}) due to fetch failure"
+              )
+              messageScheduler.schedule(
```

Comment:
I think that I am going to backtrack on creating a new PR, because I think that the `RESUBMIT_TIMEOUT` actually does still make sense.

If we go way back in DAGScheduler history (https://github.com/apache/spark/blob/branch-0.8/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala) we'll find that we had an event queue that was polled every 10 millis (`POLL_TIMEOUT`) and that fetch failures didn't produce separate resubmit tasks events, but rather we called `resubmitFailedStages` within the event polling/handling loop if 50 millis (`RESUBMIT_TIMEOUT`) had passed since the last time a `FetchFailed` was received in a `CompletionEvent`.  The handling of a `FetchFailure` didn't have any check of whether a failed stage was already in `failedStages` or any other kind of heuristic as to whether a resubmit was already scheduled for a failed stage, so the `RESUBMIT_TIMEOUT` was effectively the only thing preventing multiple resubmits from occurring as quickly as every 10 ms.

`RESUBMIT_TIMEOUT` has persisted through the multiple iterations of the DAGScheduler since then, and now it is effectively making sure that the time between `ResubmitFailedStages` events is at least 200 ms.  The time to actually complete any of the stages in `failedStages` when the `ResubmitFailedStages` event is handled doesn't really come into play, since `failedStages` is cleared within `resubmitFailedStages` and that method returns as soon as the stages that were queued up are resubmitted, not actually done re-calculating.  In other words, handling a `ResubmitFailedStages` event should be quick, and causes `failedStages` to be cleared, allowing the next `ResubmitFailedStages` event to be posted from the handling of another `FetchFailed`.  If there are the expected lot of fetch failures for a single stage, and there is no `RESUBMIT_TIMEOUT`, then it is quite likely that there will be a burst of resubmit events (and corresponding log messages) and `submitStage` calls made in rapid succession.

I think I'm inclined to keep the `RESUBMIT_TIMEOUT` throttle.